### PR TITLE
Fix cursor on emojis

### PR DESCRIPTION
**Pull request by nightwing**
- fixes cursor displaying � when placed over emoji
- fixes a bug of vim placing cursor inside emojis and other surrogate pairs
- fixes cursor bug described in the following comment https://github.com/replit/codemirror-vim/issues/38#issuecomment-1201072187

### CODE REVIEW

File: src/block-cursor.ts [hunk 1]
```diff
@@ -158,11 +158,15 @@ function measureCursor(cm: CodeMirror, view: EditorView, cursor: SelectionRange,
       domAtPos = {node: domAtPos.node.childNodes[domAtPos.offset], offset: 0};
     }
     if (!(node instanceof HTMLElement)) {
+      if (!node.parentNode) return null;
       node = node.parentNode;
     }
     let style = getComputedStyle(node as HTMLElement);
     let letter = head < view.state.doc.length && view.state.sliceDoc(head, head + 1);
     if (!letter || letter == "\n" || letter == "\r") letter = "\xa0";
+    else if ((/[\uD800-\uDC00]/.test(letter) && head < view.state.doc.length - 1)) {
```
Comment (xyc) [inline src/block-cursor.ts:167]: can we add surrogate pair comment here too?

Comment (xyc) [reply src/block-cursor.ts:167]: reading the [doc](https://learn.microsoft.com/en-us/globalization/encoding/surrogate-pairs) here, high surrogate ranges from U+D800 to U+DBFF. should this include `\uDC00`?

File: src/vim.js [hunk 2]
```diff
@@ -2953,14 +2953,28 @@ export function initVim(CodeMirror) {
...
       var vim = cm.state.vim;
       var includeLineBreak = vim.insertMode || vim.visualMode;
       var line = Math.min(Math.max(cm.firstLine(), cur.line), cm.lastLine() );
-      var maxCh = lineLength(cm, line) - 1 + !!includeLineBreak;
+      var text = cm.getLine(line);
+      var maxCh = text.length - 1 + !!includeLineBreak;
       var ch = Math.min(Math.max(0, cur.ch), maxCh);
+      // prevent cursor from entering surrogate pair
+      var charCode = text.charCodeAt(ch);
+      if (0xDC00 < charCode && charCode <0xDFFF) {
```
Comment (xyc) [inline src/vim.js:2968]: similar question to above, low surrogate ranges from U+DC00 to U+DFFF. should this range be inclusive?

File: src/vim.js [hunk 3]
```diff
@@ -2953,14 +2953,28 @@ export function initVim(CodeMirror) {
...
+        var direction = 1;
+        if (oldCur && oldCur.line == line) {
+          if (oldCur.ch > ch) {
```
Comment (xyc) [inline src/vim.js:2971]: nit: combine the conditions
```suggestion
        if (oldCur && oldCur.line == line && oldCur.ch > ch) {
```

File: src/vim.js [hunk 4]
```diff
@@ -2953,14 +2953,28 @@ export function initVim(CodeMirror) {
...
+            direction = -1;
+          }
+        }
+        ch +=direction;
+        if (ch > maxCh) ch -=2;
```
Comment (xyc) [inline src/vim.js:2976]: what does `ch -= 2` do?

File: src/vim.js [hunk 5]
```diff
@@ -2953,14 +2953,28 @@ export function initVim(CodeMirror) {
...
      * If includeLineBreak is true, then allow cur.ch == lineLength.
      */
-    function clipCursorToContent(cm, cur) {
+    function clipCursorToContent(cm, cur, oldCur) {
       var vim = cm.state.vim;
       var includeLineBreak = vim.insertMode || vim.visualMode;
       var line = Math.min(Math.max(cm.firstLine(), cur.line), cm.lastLine() );
-      var maxCh = lineLength(cm, line) - 1 + !!includeLineBreak;
+      var text = cm.getLine(line);
+      var maxCh = text.length - 1 + !!includeLineBreak;
```
Comment (xyc) [inline src/vim.js:2964]: i know this is from original code, but we should convert the types explicitly to numbers
```suggestion
      var maxCh = text.length - 1 + Number(!!includeLineBreak);
```

**Review by xyc — COMMENTED** [review]
🤗 cursor looks great! couple of nits and questions. (one thing I found is pressing `x` doesn't delete the whole emoji character but it should be fine for the scope of this PR as it happens before the change)

tagging @sergeichestakov for additional (but optional) 👀 as I don't have a lot of context.

**Review by sergeichestakov — APPROVED** [review]
Thanks yeah please resolve @xyc's comments. Otherwise, looks good to me

File: src/vim.js (hunk 2 above)
Comment (nightwing) [reply src/vim.js]: you are right, it should be.

File: src/vim.js (hunk 4 above)
Comment (nightwing) [reply src/vim.js:2982]: when cursor is moved from left side into surrogate pair the code above attempts to go to the left, as a result we may end up outside the line, in which case we have to step back

**Pull request merged**